src/plugins/intel_gpu/src/graph/impls/cpu/assign.cpp

```diff
@@ -45,22 +46,29 @@ struct assign_impl : public typed_primitive_impl<assign> {
     }
 
     event::ptr execute_impl(const std::vector<event::ptr>& events, assign_inst& instance) override {
+        if (!instance.can_be_optimized()) {
+            for (auto e : events) {
+                e->wait();
+            }
+        }
+
         auto& variable = instance.get_network().get_variable(variable_id);
+        auto &stream = instance.get_network().get_stream();
 
         OPENVINO_ASSERT(variable.get_layout() == instance.get_output_layout(),
                         "[GPU] Layout mismatch: variable layout: ", variable.get_layout().to_short_string(),
                         " assign output layout: ", instance.get_output_layout().to_short_string());
 
-        auto& stream = instance.get_network().get_stream();
-
-        for (auto e : events) {
-            e->wait();
+        if (!instance.can_be_optimized()) {
+            const auto ev_set_memory = variable.get_memory()->copy_from(stream, instance.input_memory());
+            variable.set();
+            return ev_set_memory;
         }
-
-        const auto ev_set_memory = variable.get_memory()->copy_from(stream, instance.input_memory());
+        variable.set_memory(instance.input_memory_ptr());
         variable.set();
 
-        return ev_set_memory;
+        return events.empty() ? instance.get_network().get_stream().create_user_event(true)
+                              : events[0];
```

Comment:
why events[0] instead of all events?

Reply:
applied to use `group_events` as suggestion. thanks